### PR TITLE
More tags - Link and InlineText

### PR DESCRIPTION
**Pull request by graycodes**

### CODE REVIEW

File: src/InlineText.php [hunk 1]
```diff
@@ -7,4 +7,34 @@
...
+    {
+        parent::__construct($text);
+        $this->mdTag = $markdownTag;
+    }
+
+    public function write()
+    {
+        $tag = parent::write($this->text);
+        if (!trim($tag)) return $tag;
+        $tag = trim($tag);
```
Comment (srilq) [inline src/InlineText.php]: Why do we trim whitespace here and not below where the trimmed value is used?

This has the effect that `\n\n\n A \n\n\n` becomes `A` but `\n\n\n             \n\n\n` is returned as-is. It seems arbitrary from the user's perspective that changing a single character has such an effect.

File: src/InlineText.php [hunk 2]
```diff
@@ -7,4 +7,34 @@
...
+
+        if (preg_match("/\n/", $tag)) {
+          $tag = explode("\n", $tag);
+          $tag = array_map(function ($splitTag) {
+            if (preg_match(blockRegex, $splitTag)) return $splitTag;
+            return (new InlineText($splitTag, $this->mdTag))->write();
+          }, $tag);
+          $tag = implode("\n", $tag);
+        } else {
+          $whitespace = getOuterSpace($this->text);
```
Comment (srilq) [inline src/InlineText.php]: Note: if we didn't trim above, we could use the value of `$tag` here instead of `$this->text`.

File: src/Link.php [hunk 3]
```diff
@@ -7,4 +7,44 @@
...
+
+    public function write()
+    {
+        if ($this->text == "") return self::beautify($this->text);
+
+        if (preg_match("/\n/", $this->text)) {
+            $tags = explode("\n", $this->text);
+            $tags = array_map(function($tag) {
+              if (preg_match(blockRegex, $tag)) return $tag;
+              if (!trim($tag)) return $tag;
```
Comment (srilq) [inline src/Link.php]: FYI we could use `empty(trim($tag))` if it reads better. I think this is fine as-is.

File: tests/InlineTextTest.php [hunk 4]
```diff
@@ -7,9 +7,49 @@ class InlineTextTest extends CopilotTagTest
     public function expectedWrites()
     {
         return [
+            [
+                new InlineText(),
```
Comment (srilq) [inline tests/InlineTextTest.php]: This should throw an Exception. Maybe you need to rebase this branch?

File: tests/LinkTest.php [hunk 5]
```diff
@@ -6,10 +6,76 @@ class LinkTest extends CopilotTagTest
 {
     public function expectedWrites()
     {
+        $assetMarkdown = "\n\n[#image: /photos/123ID]|||some caption|||\n";
```
Comment (srilq) [inline tests/LinkTest.php]: I prefer the idea of generating this value rather than having it inline to ensure we are using an up-to-date value. What was the rationale of not doing so here?

File: tests/LinkTest.php [hunk 6]
```diff
@@ -6,10 +6,76 @@ class LinkTest extends CopilotTagTest
...
+                new Link("some text yo $assetMarkdown", "http://li.nk"),
+                "[some text yo](http://li.nk) $assetMarkdown"
+            ],
+            [
+                new Link("$assetMarkdown some text yo", "http://li.nk"),
+                "{$assetMarkdown} [some text yo](http://li.nk)"
+            ],
+            [
+                new Link("this is $assetMarkdown some text yo", "http://li.nk"),
+                "[this is](http://li.nk) {$assetMarkdown} [some text yo](http://li.nk)"
```
Comment (srilq) [inline tests/LinkTest.php]: Are we testing for newlines in the URL at all? If we provide newlines and they are not removed, that will mess up the link generator because it splits over newlines and does a recursive call for each line, so it's important that we test this case despite it being a ridiculous input.

File: tests/LinkTest.php [hunk 7]
```diff
@@ -6,10 +6,76 @@ class LinkTest extends CopilotTagTest
...
+                new Link($assetMarkdown, "http://li.nk"),
+                "$assetMarkdown"
+            ],
+            [
+                new Link("Hello\nworld!", "http://li.nk"),
+                "[Hello](http://li.nk)\n[world!](http://li.nk)"
+            ],
+            [
+                new Link("{$assetMarkdown}{$assetMarkdown}", "http://li.nk"),
+                "{$assetMarkdown}{$assetMarkdown}"
```
Comment (srilq) [inline tests/LinkTest.php]: Am I correct in thinking would be 3 newlines in the middle of this string? We should be reducing 3 newlines to 2 in the beautify function I thought?

File: tests/LinkTest.php [hunk 8]
```diff
@@ -6,10 +6,76 @@ class LinkTest extends CopilotTagTest
...
+                new Link("Hello world!\n\n", "http://li.nk"),
+                "[Hello world!](http://li.nk)\n\n"
+            ],
+            [
+                new Link("  Hello world!\n\n", "http://li.nk"),
+                "  [Hello world!](http://li.nk)\n\n"
+            ],
+            [
+                new Link("  Hello world!\n\n", "http://li.nk"),
+                "  [Hello world!](http://li.nk)\n\n"
```
Comment (srilq) [inline tests/LinkTest.php]: What's the difference between this and the test above it?

File: src/InlineText.php [hunk 9]
```diff
@@ -7,4 +7,34 @@
  */
 class InlineText extends Text
 {
+
+    public function __construct($text = "", $markdownTag = "")
```
Comment (srilq) [inline src/InlineText.php]: There shouldn't be a default value here. (see comment further down about rebase)

File: src/CopilotTag.php [hunk 10]
```diff
@@ -1,6 +1,20 @@
 <?php
 namespace CopilotTags;
 
+
+// TODO: Put these somewhere sensible
+const blockRegex = "/\[#\w+:\s*\S+?(?:\s?(?:\w+:\s?\w+)*?)\](?:\((?:\d+(?:%|em|ex|px|pt|pc|cm|mm|in)?)(?:x?(?:\d+(?:%|em|ex|px|pt|pc|cm|mm|in)?)*)\))*(?:\|{3}[\S\s]*?\|{3})*/";
+
+function getOuterSpace($text = "") {
```
[the remark this answers is not in the document]
Comment (srilq) [reply src/CopilotTag.php]: (I could be wrong about the global scope thing, maybe it's fine because it's namespaced... I still think it should be a static function on a class).

File: src/CopilotTag.php [hunk 11]
```diff
@@ -1,6 +1,20 @@
 <?php
 namespace CopilotTags;
 
+
+// TODO: Put these somewhere sensible
+const blockRegex = "/\[#\w+:\s*\S+?(?:\s?(?:\w+:\s?\w+)*?)\](?:\((?:\d+(?:%|em|ex|px|pt|pc|cm|mm|in)?)(?:x?(?:\d+(?:%|em|ex|px|pt|pc|cm|mm|in)?)*)\))*(?:\|{3}[\S\s]*?\|{3})*/";
```
[the remark this answers is not in the document]
Comment (graycodes) [reply src/CopilotTag.php]: Cool, I'll use that one.

File: src/CopilotTag.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (graycodes) [reply src/CopilotTag.php]: > I'm wondering why this is named similarly to the outerspace library when the functionality is quite different.

Just because it gets the outer space. Wasn't meant to be similar.

> This needs to be a static function on a class. Otherwise it is probably muddying the global scope as a side-effect of autoloading, which a class library definitely should not be doing.

Cool, was wondering what to do with it, as per the comment on line 5.

File: src/InlineText.php [hunk 12]
```diff
@@ -7,4 +7,34 @@
...
+    {
+        parent::__construct($text);
+        $this->mdTag = $markdownTag;
```
[the remark this answers is not in the document]
Comment (graycodes) [reply src/InlineText.php]: Oh yeah, I just saw you'd been adding privates to other classes after I pushed this. What's the difference here? Is it just that it's automatically public otherwise?

File: src/CopilotTag.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (srilq) [reply src/CopilotTag.php]: @graycodes oh I see, I didn't see that was referring to both!

[the remark this answers is not in the document]
Comment (graycodes) [reply src/CopilotTag.php]: It's cool! Wanted to know where was sensible to put it!

File: src/InlineText.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (srilq) [reply src/InlineText.php]: @graycodes not sure, but it should be specified for readability anyway.

File: src/InlineText.php (hunk 1 above)
Comment (graycodes) [reply src/InlineText.php]: Oh, I thought this was the desired outcome?

File: src/InlineText.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (graycodes) [reply src/InlineText.php]: k, cool

File: tests/LinkTest.php (hunk 5 above)
Comment (graycodes) [reply tests/LinkTest.php]: **TL;DR: I avoid executing code from one module to test another, just in case they're both equally wrong and your tests still pass.**

I just wanted to be more explicit about what the value is of asset markdown. When I was looking through the tests in Tugboat for this, I just wanted to know what this value actually was. It's not a big deal, but I feel it's more explicit.

How often is the markdown spec going to change?

You're right in that if it changes we'll also have to update it here, and that it could end up being stale. However, I can foresee some situation where we're doing `$assetMarkdown = new Asset('a', 'b', 'c');` and (for some reason) `$assetMarkdown` gets set to null or an empty string or something (bad import, old code, edge cases, typos, blah) and all of our tests are passing but they're not testing what we think they are.

Again this isn't super likely either but there felt like code smell. I generally try and avoid executing code from one module to test another, just in case they're both equally wrong and your tests still pass.

File: tests/LinkTest.php (hunk 8 above)
Comment (graycodes) [reply tests/LinkTest.php]: Check's if people are really reading the PRs ;)

File: src/InlineText.php (hunk 1 above)
Comment (srilq) [reply src/InlineText.php]: @graycodes I don't think this was how it worked previously. This way messes up the fairly common case of having leading/trailing whitespace in XML tags:
```xml
<strong>my strong
text </strong>is strong
```
becomes:
```md
**my strong**
**text**is strong
```
instead of:
```md
**my strong**
**text** is strong
```

Comment (graycodes) [reply src/InlineText.php]: Oooooh k

Comment (srilq) [reply src/InlineText.php]: @graycodes I'm wondering if it would be easier to just re-write the `outerspace` API in PHP and do a more direct translation of the Tugboat code to ensure we don't have regressions like this being introduced. We've been through the process of fixing the many corner cases and simplifying logic already with the CopilotTags rewrite so I don't see the value in changing the logic again now.

Comment (srilq) [reply src/InlineText.php]: From what I've seen it's mostly fine, there are just a few things that have changed that are concerning me!

File: tests/LinkTest.php (hunk 5 above)
Comment (srilq) [reply tests/LinkTest.php]: @graycodes I wasn't concerned about the spec changing, more concerned about the whitespace being added in `Embed::write` changing. We are adding 2 leading newlines and 1 trailing at the moment, but that might not always be the case, and we just have to remember to update this string literal if we change the behaviour of `Embed`.

I see what you mean though. I don't mind either way really.

File: src/InlineText.php (hunk 1 above)
Comment (graycodes) [reply src/InlineText.php]: Wasn't an intentional change of logic. I thought I had all the tests covered but clearly there's some gaps

Comment (srilq) [reply src/InlineText.php]: @graycodes yeah, gaps in tests was definitely a big issue that I'm hoping we can remedy in the PHP version.. I realise it makes it harder to validate that the PHP version works the same way as before. 😅

File: tests/LinkTest.php (hunk 6 above)
Comment (graycodes) [reply tests/LinkTest.php]: There is now a test for newlines in links

File: tests/LinkTest.php (hunk 5 above)
Comment (srilq) [reply tests/LinkTest.php]: @graycodes can I suggest a compromise - we generate the input but write the expected output as a string literal?

Comment (graycodes) [reply tests/LinkTest.php]: I guess I'll update once we create the asset tag

Comment (srilq) [reply tests/LinkTest.php]: @graycodes I don’t see the point of `Asset` and it’s not in the spec. `Embed` will do.

Comment (graycodes) [reply tests/LinkTest.php]: ah okay
